File: Server/src/main/core/cache/def/impl/DataMap.java
package core.cache.def.impl;

import core.cache.Cache;
import core.cache.misc.buffer.ByteBufferUtils;
import core.tools.CP1252;
import core.tools.Log;

import java.nio.ByteBuffer;
import java.util.HashMap;
import java.util.Map;

import static core.api.ContentAPIKt.log;

/**
 * Data map.
 */
public class DataMap {

    private static final Map<Integer, DataMap> definitions = new HashMap<>();

    private final int id;

    /**
     * The Key type.
     */
    public char keyType = '?';

    /**
     * The Value type.
     */
    public char valueType = '?';

    /**
     * The Default string.
     */
    public String defaultString;

    /**
     * The Default int.
     */
    public int defaultInt;

    /**
     * The Data store.
     */
    public HashMap<Integer, Object> dataStore = new HashMap<>();

    /**
     * Instantiates a new Data map.
     *
     * @param id the id
     */
    public DataMap(int id) {
        this.id = id;
    }

    /**
     * Gets int.
     *
     * @param key the key
     * @return the int
     */
    public int getInt(int key) {
        if (!dataStore.containsKey(key)) {
            log(this.getClass(), Log.ERR, "Invalid value passed for key: " + key + " map: " + id);
            return -1;
        }
        return (int) dataStore.get(key);
    }

    /**
     * Gets string.
     *
     * @param key the key
     * @return the string
     */
    public String getString(int key) {
        return (String) dataStore.get(key);
    }

    @Override
    public String toString() {
        StringBuilder builder = new StringBuilder();
        return "DataMapDefinition{" +
            "id=" + id +
            ", keyType=" + keyType +
            ", valueType=" + (valueType == 'K' ? "Normal" : valueType == 'J' ? "Struct Pointer" : "Unknown") +
            ", defaultString='" + defaultString + '\'' +
            ", defaultInt=" + defaultInt +
            ", dataStore=" + dataStore +
            '}' + "\n";
    }

    /**
     * Get data map.
     *
     * @param id the id
     * @return the data map
     */
    public static DataMap get(int id) {
        DataMap def = definitions.get(id);
        if (def != null) {
            return def;
        }
        byte[] data = Cache.getIndexes()[17].getFileData(id >>> 8, id & 0xFF);
        def = parse(id, data);
        definitions.put(id, def);
        return def;
    }

    /**
     * Parse data map.
     *
     * @param id   the id
     * @param data the data
     * @return the data map
     */
    public static DataMap parse(int id, byte[] data) {
        DataMap def = new DataMap(id);
        if (data != null) {
            ByteBuffer buffer = ByteBuffer.wrap(data);
            int opcode;

            while ((opcode = buffer.get() & 0xFF) != 0) {

                if (opcode == 1) {
                    def.keyType = CP1252.decode(buffer.get());
                } else if (opcode == 2) {
                    def.valueType = CP1252.decode(buffer.get());
                } else if (opcode == 3) {
                    def.defaultString = ByteBufferUtils.getString(buffer);
                } else if (opcode == 4) {
                    def.defaultInt = buffer.getInt();
                } else if (opcode == 5 || opcode == 6) {
                    int size = buffer.getShort() & 0xFFFF;

                    for (int i = 0; i < size; i++) {
                        int key = buffer.getInt();

                        Object value;
                        if (opcode == 5) {
                            value = ByteBufferUtils.getString(buffer);
                        } else {
                            value = buffer.getInt();
                        }
                        def.dataStore.put(key, value);
                    }
                }
            }
        }
        return def;
    }

    /**
     * Gets id.
     *
     * @return the id
     */
    public int getId() {
        return id;
    }
}

File: Emulator/src/main/core/cache/def/impl/DataMap.kt
package core.cache.def.impl

import core.api.log
import core.cache.Cache
import core.cache.misc.buffer.ByteBufferUtils
import core.tools.CP1252
import core.tools.Log
import core.tools.SystemLogger
import java.nio.ByteBuffer

class DataMap(private val id: Int) {

    /**
     * The enum id.
     */
    var keyType: Char = '?'
    var valueType: Char = '?'
    var defaultString: String? = null
    var defaultInt: Int = 0
    var dataStore: HashMap<Int, Any> = HashMap()

    companion object {
        /**
         * The config definitions mapping.
         */
        private val definitions: MutableMap<Int, DataMap> = HashMap()

        /**
         * Get data map.
         *
         * @param id the id.
         * @return the data map.
         */
        fun get(id: Int): DataMap {
            val def = definitions[id]
            if (def != null) {
                return def
            }
            val data = Cache.getIndexes()[17].getFileData(id shr 8, id and 0xFF)
            val parsedDef = parse(id, data)
            definitions[id] = parsedDef
            return parsedDef
        }

        /**
         * Parse data map.
         *
         * @param id   the id.
         * @param data the data.
         * @return the data map.
         */
        fun parse(id: Int, data: ByteArray?): DataMap {
            val def = DataMap(id)
            if (data != null) {
                val buffer = ByteBuffer.wrap(data)
                var opcode: Int

                while (buffer.get().toInt() and 0xFF != 0) {
                    opcode = buffer.get().toInt() and 0xFF

                    when (opcode) {
                        1 -> def.keyType = CP1252.decode(buffer.get())
                        2 -> def.valueType = CP1252.decode(buffer.get())
                        3 -> def.defaultString = ByteBufferUtils.getString(buffer)
                        4 -> def.defaultInt = buffer.int
                        5, 6 -> {
                            val size = buffer.short.toInt() and 0xFFFF
                            for (i in 0 until size) {
                                val key = buffer.int
                                val value: Any = if (opcode == 5) {
                                    ByteBufferUtils.getString(buffer)!!
                                } else {
                                    buffer.int
                                }
                                def.dataStore[key] = value
                            }
                        }
                    }
                }
            }
            return def
        }
    }

    override fun toString(): String {
        return "DataMapDefinition{" +
                "id=$id" +
                ", keyType=$keyType" +
                ", valueType=${if (valueType == 'K') "Normal" else if (valueType == 'J') "Struct Pointer" else "Unknown"}" +
                ", defaultString='$defaultString'" +
                ", defaultInt=$defaultInt" +
                ", dataStore=$dataStore" +
                '}' + "\n"
    }

    fun getString(key: Int): String? {
        return dataStore[key] as String?
    }

    fun getInt(key: Int): Int {
        if (!dataStore.containsKey(key)) {
            log(this::class.java, Log.ERR, "Invalid value passed for key: $key map: $id")
            return -1
        }
        return dataStore[key] as Int
    }

    fun getId(): Int {
        return id
    }
}
